tekton/resources/cd/eventlistener.yaml

```diff
@@ -50,3 +50,14 @@ spec:
         - name: helm-details
       template:
         name: deploy-helm-chart
+    - name: tekton
+      interceptors:
+        - cel:
+            filter: >-
+              'tekton' in body
```

Comment:
i think this is a continuation of the conversation we started in https://github.com/tektoncd/plumbing/pull/270/files#r392387792 but i still dont quite understand what this filter means - i understand wanting to have one event listener that works for multiple triggers, but what does it mean to have just the word "tekton" in the body, why is that enough to uniquely identify this? or `folder` in the other comment?

(also in any case it might be worth some comments explaining?)

Reply:
This filters create some kind of "endpoint" for the different CD services within this event listener. We control the body that it is sent.
All CD services have a few shared params, which is also reflected in the structure of the bindings. Params that are specific to a CD service are wrapped within a key in the `body` map. 

You can find the following in the [base folders under cronjobs](https://github.com/tektoncd/plumbing/tree/master/tekton/cronjobs):

For folders:
```
                {
                  "git": {
                    "repository": "$GIT_REPOSITORY",
                    "revision": "$GIT_REVISION"
                  },
                  "namespace": "$NAMESPACE",
                  "cluster-resource": "$CLUSTER_RESOURCE",
                  "folder": {
                      "path": "$FOLDER_PATH",
                      "description": "$FOLDER_DESCRIPTION",
                      "method": "$FOLDER_METHOD",
                      "is-overlay": "$FOLDER_OVERLAY"
                  }
                }
```

For config maps:
```
                {
                  "git": {
                    "repository": "$GIT_REPOSITORY",
                    "revision": "$GIT_REVISION"
                  },
                  "namespace": "$NAMESPACE",
                  "cluster-resource": "$CLUSTER_RESOURCE",
                  "configmap": {
                      "name": "$CONFIGMAP_NAME",
                      "key": "$CONFIGMAP_KEY",
                      "description": "$CONFIGMAP_DESCRIPTION",
                      "path": "$CONFIG_PATH"
                  }
                }
```

For helm:
```
                {
                  "namespace": "$NAMESPACE",
                  "cluster-resource": "$CLUSTER_RESOURCE",
                  "helm": {
                      "name": "$CHART_NAME",
                      "version": "$CHART_VERSION",
                      "repo": "$CHART_REPO",
                      "params": "$CHART_PARAMS",
                      "description": "$CHART_DESCRIPTION",
                      "pre-deploy-resources": "$PRE_DEPLOY_RESOURCES"
                  }
                }
```

For tekton services:
```
                {
                  "namespace": "$NAMESPACE",
                  "cluster-resource": "$CLUSTER_RESOURCE",
                  "tekton": {
                    "project": "$TEKTON_PROJECT",
                    "version": "$TEKTON_VERSION",
                    "environment": "$TEKTON_CLUSTER",
                    "bucket": "$RELEASE_BUCKET",
                    "file": "$RELEASE_FILE",
                    "extra-path": "$RELEASE_EXTRA_PATH"
                  },
                  "plumbing": {
                    "repository": "$PLUMBING_REPOSITORY",
                    "revision": "$PLUMBING_REVISION"
                  }
                }
```

Reply:
What I am understanding is that this is like a "command" based interface, kinda like chatops but generated from crons, where we have one endpoint listening, and based on the body of the request, we're determining what command/triggertemplate to invoke, is that right?

> 'tekton' in body

My understanding of this line is that if the word "tekton" appears _anywhere_ in the body (or does the `in` operator just look at keys? i cant tell from [the docs](https://github.com/google/cel-spec/blob/master/doc/langdef.md)),  then the tekton services template will be invoked, is that right?

What do you think about making this relationship a bit more explicit, e.g. by including something like a `type`, `command`, or `triggertemplate` field? e.g.:'

```json
               {
                  "namespace": "$NAMESPACE",
                  "cluster-resource": "$CLUSTER_RESOURCE",
                  "trigger-template": "deploy-tekton-release",
                  "tekton": {
                    ...
                  },
                  "plumbing": {
                     ...
                  }
                }
```

We could even make it super generic and have a "params" field, e.g.

```json
               {
                  "namespace": "$NAMESPACE",
                  "cluster-resource": "$CLUSTER_RESOURCE",
                  "trigger-template": "deploy-tekton-release",
                  "params": {
                      "tekton": {
                        ...
                      },
                      "plumbing": {
                         ...
                      }
                  }
                }
```

I'm not sure if it's better but I do think it would make it pretty clear what was going on

Reply:
'tekton' in body in CEL syntax means that the string "tekton" is a key in the map "body"

Reply:
> What I am understanding is that this is like a "command" based interface, kinda like chatops but generated from crons, where we have one endpoint listening, and based on the body of the request, we're determining what command/triggertemplate to invoke, is that right?
> 
> > 'tekton' in body
> 
> My understanding of this line is that if the word "tekton" appears _anywhere_ in the body (or does the `in` operator just look at keys? i cant tell from [the docs](https://github.com/google/cel-spec/blob/master/doc/langdef.md)), then the tekton services template will be invoked, is that right?
> 
> What do you think about making this relationship a bit more explicit, e.g. by including something like a `type`, `command`, or `triggertemplate` field? e.g.:'
> 
> ```json
>                {
>                   "namespace": "$NAMESPACE",
>                   "cluster-resource": "$CLUSTER_RESOURCE",
>                   "trigger-template": "deploy-tekton-release",
>                   "tekton": { },
>                   "plumbing": { }
>                 }
> ```
> 
> We could even make it super generic and have a "params" field, e.g.
> 
> ```json
>                {
>                   "namespace": "$NAMESPACE",
>                   "cluster-resource": "$CLUSTER_RESOURCE",
>                   "trigger-template": "deploy-tekton-release",
>                   "params": {
>                       "tekton": { },
>                       "plumbing": { }
>                   }
>                 }
> ```
> 
> I'm not sure if it's better but I do think it would make it pretty clear what was going on

Sounds like a good idea - the CEL filter won't get much more complicated.

Reply:
I'm working towards something like this:
```
{
                  "trigger-template": "configmaps",
                  "params": {
                    "git": {
                      "repository": "$GIT_REPOSITORY",
                      "revision": "$GIT_REVISION"
                    },
                    "target": {
                      "namespace": "$NAMESPACE",
                      "cluster-resource": "$CLUSTER_RESOURCE",
                    }
                    "configmap": {
                        "description": "$CONFIGMAP_DESCRIPTION",
                        "path": "$CONFIG_PATH"
                        "name": "$CONFIGMAP_NAME",
                        "key": "$CONFIGMAP_KEY"
                    },
                  }
                }
```
Every key in the params map corresponds to a different binding.

Reply:
i like it!!! :D

Reply:
> 'tekton' in body in CEL syntax means that the string "tekton" is a key in the map "body"

aahhhh kk that makes more sense! thanks for explaining